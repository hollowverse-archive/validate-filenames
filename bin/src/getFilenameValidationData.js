'use strict';

Object.defineProperty(exports, '__esModule', { value: true });
var minimatch = require('minimatch');
var validateFilenameAgainstRule_1 = require('./validateFilenameAgainstRule');
function filenameMatchesPattern(filename, pattern) {
  return minimatch(filename, pattern, { matchBase: true });
}
function filenameMatchesAnyPattern(filename, patterns) {
  return patterns.some(function(pattern) {
    return filenameMatchesPattern(filename, pattern);
  });
}
function getLastApplicableRule(filename, rules) {
  var lastApplicableRule = void 0;
  var _iteratorNormalCompletion = true;
  var _didIteratorError = false;
  var _iteratorError = undefined;

  try {
    for (
      var _iterator = rules[Symbol.iterator](), _step;
      !(_iteratorNormalCompletion = (_step = _iterator.next()).done);
      _iteratorNormalCompletion = true
    ) {
      var rule = _step.value;

      if (filenameMatchesAnyPattern(filename, rule.patterns)) {
        lastApplicableRule = rule;
      }
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator.return) {
        _iterator.return();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }

  return lastApplicableRule;
}
function getFilenameValidationData(filenames, rules) {
  return filenames.map(function(filename) {
    var result = {
      valid: true,
      invalidComponents: [],
    };
    var lastApplicableRule = getLastApplicableRule(filename, rules);
    if (lastApplicableRule) {
      result = validateFilenameAgainstRule_1.validateFilenameAgainstRule(
        filename,
        lastApplicableRule,
      );
    }
    return [filename, result];
  });
}
exports.getFilenameValidationData = getFilenameValidationData;
