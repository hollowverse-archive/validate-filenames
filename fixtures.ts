export const filenames = [
  // dot file
  '.dockerignore',

  // dot folder
  '.elasticbeanstalk/config.yml',

  // Capital letter, no extension
  'Dockerfile',

  // Capital letter, no extension, with dash
  'Dockerfile-dev',

  // All caps with extension
  'README.md',

  // All lower case
  'build.js',
  'commonconfig.js',
  'database.rules.json',

  // Snake case
  'docker-compose.yml',

  // All lower case (secrets/ path)
  'secrets/sumo.json.d.ts',
  'secrets/sumo.json.enc',

  // Dot file (src/app/)
  'src/app/.babelrc',

  // Begin with underscore (src/app/)
  'src/app/_variables.scss',

  // camelCase (src/app/)
  'src/app/clientEntry.ts',

  // Pascal case (src/app/components/)
  'src/app/components/App/App.global.scss',
  'src/app/components/App/App.tsx',

  // Pascal case (src/app/pages)
  'src/app/pages/NavBar/NavBar.module.scss',
  'src/app/pages/NavBar/NavBar.tsx',

  // Snake case (src/typings/)
  'src/typings/shrink-ray.d.ts',
  'src/typings/webpack-hot-server-middleware.d.ts',
];
