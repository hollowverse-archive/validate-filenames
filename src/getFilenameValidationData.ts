import * as minimatch from 'minimatch';
import { validateFilenameAgainstRule } from './validateFilenameAgainstRule';
import { Rule, FilenameValidationResult } from './types';

function filenameMatchesPattern(filename: string, pattern: string) {
  return minimatch(filename, pattern, { matchBase: true });
}

function filenameMatchesAnyPattern(filename: string, patterns: string[]) {
  return patterns.some(pattern => {
    return filenameMatchesPattern(filename, pattern);
  });
}

function getLastApplicableRule(filename: string, rules: Rule[]) {
  let lastApplicableRule;

  for (const rule of rules) {
    if (filenameMatchesAnyPattern(filename, rule.patterns)) {
      lastApplicableRule = rule;
    }
  }

  return lastApplicableRule;
}

export function getFilenameValidationData(filenames: string[], rules: Rule[]) {
  return filenames.map(filename => {
    let result: FilenameValidationResult = {
      valid: true,
      invalidComponents: [],
    };
    const lastApplicableRule = getLastApplicableRule(filename, rules);

    if (lastApplicableRule) {
      result = validateFilenameAgainstRule(filename, lastApplicableRule);
    }

    return [filename, result] as [string, FilenameValidationResult];
  });
}
