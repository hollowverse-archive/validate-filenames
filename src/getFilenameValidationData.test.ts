import { getFilenameValidationData } from './getFilenameValidationData';
import { filenames } from '../fixtures';

describe('getFilenameValidationData', () => {
  it('returns valid = true when filename is valid', () => {
    expect(
      getFilenameValidationData(
        ['src/app/hello.ts', 'src/app/helloWorld.ts'],
        [
          {
            validation: 'camelCase',
            patterns: ['**/*'],
          },
        ],
      ),
    ).toEqual([
      ['src/app/hello.ts', { valid: true, invalidComponents: [] }],
      ['src/app/helloWorld.ts', { valid: true, invalidComponents: [] }],
    ]);
  });

  it('returns valid = false and indexes of invalid components when filename is invalid', () => {
    expect(
      getFilenameValidationData(
        ['src/app/hello-world.ts'],
        [
          {
            validation: 'camelCase',
            patterns: ['**/*'],
          },
        ],
      ),
    ).toEqual([
      ['src/app/hello-world.ts', { valid: false, invalidComponents: [2] }],
    ]);
  });

  it('understands "camelCase"', () => {
    expect(
      getFilenameValidationData(
        ['src/app/helloWorld.ts'],
        [
          {
            validation: 'camelCase',
            patterns: ['**/*'],
          },
        ],
      ),
    ).toEqual([
      ['src/app/helloWorld.ts', { valid: true, invalidComponents: [] }],
    ]);

    expect(
      getFilenameValidationData(
        ['src/app/HelloWorld.ts'],
        [
          {
            validation: 'camelCase',
            patterns: ['**/*'],
          },
        ],
      ),
    ).toEqual([
      ['src/app/HelloWorld.ts', { valid: false, invalidComponents: [2] }],
    ]);
  });

  it('understands "PascalCase"', () => {
    expect(
      getFilenameValidationData(
        ['HelloWorld.ts'],
        [
          {
            validation: 'PascalCase',
            patterns: ['**/*'],
          },
        ],
      ),
    ).toEqual([['HelloWorld.ts', { valid: true, invalidComponents: [] }]]);

    expect(
      getFilenameValidationData(
        ['helloWorld.ts'],
        [
          {
            validation: 'PascalCase',
            patterns: ['**/*'],
          },
        ],
      ),
    ).toEqual([['helloWorld.ts', { valid: false, invalidComponents: [0] }]]);
  });

  it('does not validate components that appear in the pattern', () => {
    expect(
      getFilenameValidationData(
        ['src/app/HelloWorld.ts'],
        [
          {
            validation: 'PascalCase',
            patterns: ['src/app/**/*'],
          },
        ],
      ),
    ).toEqual([
      ['src/app/HelloWorld.ts', { valid: true, invalidComponents: [] }],
    ]);

    expect(
      getFilenameValidationData(
        ['src/app/helloWorld.ts'],
        [
          {
            validation: 'PascalCase',
            patterns: ['src/app/**/*'],
          },
        ],
      ),
    ).toEqual([
      ['src/app/helloWorld.ts', { valid: false, invalidComponents: [2] }],
    ]);
  });

  it('consolidates pattern components correctly when it ignores filename components in the pattern', () => {
    expect(
      getFilenameValidationData(
        ['src/app/pages/NavBar/NavBar.module.scss'],
        [
          {
            validation: 'PascalCase',
            patterns: ['src/app/components/**/*', 'src/app/pages/**/*'],
          },
        ],
      ),
    ).toEqual([
      [
        'src/app/pages/NavBar/NavBar.module.scss',
        { valid: true, invalidComponents: [] },
      ],
    ]);
  });

  it('complains about camelCase or PascalCase that have two consecutive uppercase letters', () => {
    expect(
      getFilenameValidationData(
        ['src/app/helloWOrld.ts'],
        [
          {
            validation: 'camelCase',
            patterns: ['**/*'],
          },
        ],
      ),
    ).toEqual([
      ['src/app/helloWOrld.ts', { valid: false, invalidComponents: [2] }],
    ]);

    expect(
      getFilenameValidationData(
        ['src/app/HelloWOrld.ts'],
        [
          {
            validation: 'PascalCase',
            patterns: ['src/app/*'],
          },
        ],
      ),
    ).toEqual([
      ['src/app/HelloWOrld.ts', { valid: false, invalidComponents: [2] }],
    ]);
  });

  it('works for `camelCase` alone', () => {
    expect(
      getFilenameValidationData(filenames, [
        {
          validation: 'camelCase',
          patterns: ['**/*'],
        },
      ]),
    ).toMatchSnapshot();
  });

  it('works when we ignore certain patterns', () => {
    expect(
      getFilenameValidationData(filenames, [
        {
          validation: 'camelCase',
          patterns: ['**/*'],
        },

        {
          validation: 'ignore',
          patterns: [
            '*/**/typings/*',
            '_*.scss',
            'Dockerfile*',
            'docker-compose.yml',
            '**/LICENSE.md',
            '**/README.md',
          ],
        },
      ]),
    ).toMatchSnapshot();
  });

  it('works when we specify different validation rules for different paths', () => {
    expect(
      getFilenameValidationData(filenames, [
        {
          validation: 'camelCase',
          patterns: ['**/*'],
        },

        {
          validation: 'PascalCase',
          patterns: ['src/app/components/**/*.tsx'],
        },

        {
          validation: 'ignore',
          patterns: [
            '*/**/typings/*',
            '_*.scss',
            'Dockerfile*',
            'docker-compose.yml',
            '**/LICENSE.md',
            '**/README.md',
          ],
        },
      ]),
    ).toMatchSnapshot();
  });

  it('works when we fine-tune the rules for the filenames so everything is valid', () => {
    expect(
      getFilenameValidationData(filenames, [
        {
          validation: 'camelCase',
          patterns: ['**/*'],
        },

        {
          validation: 'PascalCase',
          patterns: ['src/app/components/**/*', 'src/app/pages/**/*'],
        },

        {
          validation: 'ignore',
          patterns: [
            '*/**/typings/*',
            '_*.scss',
            'Dockerfile*',
            'docker-compose.yml',
            '**/LICENSE.md',
            '**/README.md',
          ],
        },
      ]),
    ).toMatchSnapshot();
  });
});
