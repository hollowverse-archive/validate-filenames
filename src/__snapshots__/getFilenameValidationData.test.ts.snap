// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFilenameValidationData works for \`camelCase\` alone 1`] = `
Array [
  Array [
    ".dockerignore",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    ".elasticbeanstalk/config.yml",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "Dockerfile",
    Object {
      "invalidComponents": Array [
        0,
      ],
      "valid": false,
    },
  ],
  Array [
    "Dockerfile-dev",
    Object {
      "invalidComponents": Array [
        0,
      ],
      "valid": false,
    },
  ],
  Array [
    "README.md",
    Object {
      "invalidComponents": Array [
        0,
      ],
      "valid": false,
    },
  ],
  Array [
    "build.js",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "commonconfig.js",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "database.rules.json",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "docker-compose.yml",
    Object {
      "invalidComponents": Array [
        0,
      ],
      "valid": false,
    },
  ],
  Array [
    "secrets/sumo.json.d.ts",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "secrets/sumo.json.enc",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/app/.babelrc",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/app/_variables.scss",
    Object {
      "invalidComponents": Array [
        2,
      ],
      "valid": false,
    },
  ],
  Array [
    "src/app/clientEntry.ts",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/app/components/App/App.global.scss",
    Object {
      "invalidComponents": Array [
        3,
        4,
      ],
      "valid": false,
    },
  ],
  Array [
    "src/app/components/App/App.tsx",
    Object {
      "invalidComponents": Array [
        3,
        4,
      ],
      "valid": false,
    },
  ],
  Array [
    "src/app/pages/NavBar/NavBar.module.scss",
    Object {
      "invalidComponents": Array [
        3,
        4,
      ],
      "valid": false,
    },
  ],
  Array [
    "src/app/pages/NavBar/NavBar.tsx",
    Object {
      "invalidComponents": Array [
        3,
        4,
      ],
      "valid": false,
    },
  ],
  Array [
    "src/typings/shrink-ray.d.ts",
    Object {
      "invalidComponents": Array [
        2,
      ],
      "valid": false,
    },
  ],
  Array [
    "src/typings/webpack-hot-server-middleware.d.ts",
    Object {
      "invalidComponents": Array [
        2,
      ],
      "valid": false,
    },
  ],
]
`;

exports[`getFilenameValidationData works when we fine-tune the rules for the filenames so everything is valid 1`] = `
Array [
  Array [
    ".dockerignore",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    ".elasticbeanstalk/config.yml",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "Dockerfile",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "Dockerfile-dev",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "README.md",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "build.js",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "commonconfig.js",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "database.rules.json",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "docker-compose.yml",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "secrets/sumo.json.d.ts",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "secrets/sumo.json.enc",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/app/.babelrc",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/app/_variables.scss",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/app/clientEntry.ts",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/app/components/App/App.global.scss",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/app/components/App/App.tsx",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/app/pages/NavBar/NavBar.module.scss",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/app/pages/NavBar/NavBar.tsx",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/typings/shrink-ray.d.ts",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/typings/webpack-hot-server-middleware.d.ts",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
]
`;

exports[`getFilenameValidationData works when we ignore certain patterns 1`] = `
Array [
  Array [
    ".dockerignore",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    ".elasticbeanstalk/config.yml",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "Dockerfile",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "Dockerfile-dev",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "README.md",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "build.js",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "commonconfig.js",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "database.rules.json",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "docker-compose.yml",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "secrets/sumo.json.d.ts",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "secrets/sumo.json.enc",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/app/.babelrc",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/app/_variables.scss",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/app/clientEntry.ts",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/app/components/App/App.global.scss",
    Object {
      "invalidComponents": Array [
        3,
        4,
      ],
      "valid": false,
    },
  ],
  Array [
    "src/app/components/App/App.tsx",
    Object {
      "invalidComponents": Array [
        3,
        4,
      ],
      "valid": false,
    },
  ],
  Array [
    "src/app/pages/NavBar/NavBar.module.scss",
    Object {
      "invalidComponents": Array [
        3,
        4,
      ],
      "valid": false,
    },
  ],
  Array [
    "src/app/pages/NavBar/NavBar.tsx",
    Object {
      "invalidComponents": Array [
        3,
        4,
      ],
      "valid": false,
    },
  ],
  Array [
    "src/typings/shrink-ray.d.ts",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/typings/webpack-hot-server-middleware.d.ts",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
]
`;

exports[`getFilenameValidationData works when we specify different validation rules for different paths 1`] = `
Array [
  Array [
    ".dockerignore",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    ".elasticbeanstalk/config.yml",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "Dockerfile",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "Dockerfile-dev",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "README.md",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "build.js",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "commonconfig.js",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "database.rules.json",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "docker-compose.yml",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "secrets/sumo.json.d.ts",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "secrets/sumo.json.enc",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/app/.babelrc",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/app/_variables.scss",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/app/clientEntry.ts",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/app/components/App/App.global.scss",
    Object {
      "invalidComponents": Array [
        3,
        4,
      ],
      "valid": false,
    },
  ],
  Array [
    "src/app/components/App/App.tsx",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/app/pages/NavBar/NavBar.module.scss",
    Object {
      "invalidComponents": Array [
        3,
        4,
      ],
      "valid": false,
    },
  ],
  Array [
    "src/app/pages/NavBar/NavBar.tsx",
    Object {
      "invalidComponents": Array [
        3,
        4,
      ],
      "valid": false,
    },
  ],
  Array [
    "src/typings/shrink-ray.d.ts",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
  Array [
    "src/typings/webpack-hot-server-middleware.d.ts",
    Object {
      "invalidComponents": Array [],
      "valid": true,
    },
  ],
]
`;
